components/validate/ajv/index.mjs

```diff
@@ -3,15 +3,17 @@ import Treeify from "treeify";
 import AjvErrorTree from "ajv-error-tree";
 import { schema } from "../../../dist/schema.mjs";
 
-const {ownKeys} = Reflect;
+const { ownKeys } = Reflect;
 const _Map = Map;
 const { asTree } = Treeify;
 
+const expected_extra_properties = ["appmap_dir"];
```

Comment:
appmap_dir is used by the agent too, at least implicitly, is it not?

Reply:
I think there’s something I don’t understand because `appmap_dir: true` doesn’t make sense to me. It’s a directory name.

Reply:
`true` is just a schema that satisfies everything. And I'm using `output.directory` instead of `appmap_dir`.

Reply:
OK. Generally we do not want people configuring custom output directories, because then other tools like index, scanner, VSCode extension don't know where to look for AppMaps - unless more additional options are provided to those tools, which is confusing to the user. Really, we just want each language/platform to have an output directory that's specified by `appmap_dir` in the _appmap.yml_, and not deviate from that unless there is a very particular and good reason. Otherwise it leads to broken user experience downstream of the agent.

Reply:
@lachrist are you saying that `language` is a field in _appmap.yml_? Or are you talking about metadata in the appmap document? This is getting really confusing.

What I'm trying to do here is to get all the agents to recognize two properties in _appmap.yml_ : `language` and `appmap_dir`.

Reply:
I just want the schema to recognize: `language: string (= javascript)` and `appmap_dir: string (= tmp/appmap)`. Can you make that happen? Is there some reason not  to do this? It will make life much simpler for the downstream tools like the indexer, scanner, VSCode extension, because then they will know (a) what language is being mapped (even if there are no AppMaps yet) (b) where the AppMaps are located

Reply:
I understand the importance of `appmap_dir`. I was not aware that it was standardized across the agents. I can easily make the agent conform to this requirement but it will introduce a breaking change and bump the agent to a new major version.

At the moment the agent recognize the `language` configuration field (ie in `appmap.yml`) as a combination of `name` and `version`. For instance: `"javascript"`, `"javascript@latest"`, `"ecmascript@5"` or `{name:"js", version:"2020"}`. It is currently used in two places:

1) Populate `metadata.language.name` in generated appmaps.
2) Pass the version to the `acorn` parser.

Passing the version to acorn is a nice to have. We could always pass `"latest"` and it should just work in theory. So I think we could drop support for configuring language version. This will also introduce a breaking change.

What do you think? Do you want me to implement these two breaking changes?